pkg/agent/controller/supportbundlecollection/support_bundle_controller.go

```diff
@@ -0,0 +1,417 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package supportbundlecollection
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"path"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/sftp"
+	"github.com/spf13/afero"
+	"golang.org/x/crypto/ssh"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/exec"
+
+	"antrea.io/antrea/pkg/agent"
+	"antrea.io/antrea/pkg/agent/config"
+	agentquerier "antrea.io/antrea/pkg/agent/querier"
+	"antrea.io/antrea/pkg/apis/controlplane"
+	cpv1b2 "antrea.io/antrea/pkg/apis/controlplane/v1beta2"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	"antrea.io/antrea/pkg/ovs/ovsctl"
+	"antrea.io/antrea/pkg/querier"
+	"antrea.io/antrea/pkg/support"
+	"antrea.io/antrea/pkg/util/compress"
+)
+
+type ProtocolType string
+
+const (
+	SFTP ProtocolType = "SFTP"
+
+	controllerName = "SupportBundleCollectionController"
+	defaultWorkers = 1
+
+	UPLOAD_TO_FILESERVER_TRIES       = 5
+	UPLOAD_TO_FILESERVER_RETRY_DELAY = 5 * time.Second
+)
+
+var (
+	emptyWatch      = watch.NewEmptyWatch()
+	defaultFS       = afero.NewOsFs()
+	defaultExecutor = exec.New()
+	// Declared as variable for testing.
+	newAgentDumper = support.NewAgentDumper
+
+	getUploader          = getUploaderByProtocol
+	sftpUploaderInstance *sftpUploader
+)
+
+type SupportBundleController struct {
+	nodeName           string
+	nodeType           config.NodeType
+	namespace          string
+	antreaClientGetter agent.AntreaClientProvider
+	queue              *workqueue.Type
+	ovsCtlClient       ovsctl.OVSCtlClient
+	aq                 agentquerier.AgentQuerier
+	npq                querier.AgentNetworkPolicyInfoQuerier
+	v4Enabled          bool
+	v6Enabled          bool
+	lastCondition      v1alpha1.SupportBundleCollectionConditionType
+	conditionLock      sync.RWMutex
+}
+
+func NewSupportBundleController(nodeName string, nodeType config.NodeType, namespace string, antreaClientGetter agent.AntreaClientProvider, ovsCtlClient ovsctl.OVSCtlClient,
+	aq agentquerier.AgentQuerier, npq querier.AgentNetworkPolicyInfoQuerier, v4Enabled, v6Enabled bool) *SupportBundleController {
+	c := &SupportBundleController{
+		nodeName:           nodeName,
+		nodeType:           nodeType,
+		antreaClientGetter: antreaClientGetter,
+		queue:              workqueue.NewNamed("supportbundle"),
+		ovsCtlClient:       ovsCtlClient,
+		aq:                 aq,
+		npq:                npq,
+		v4Enabled:          v4Enabled,
+		v6Enabled:          v6Enabled,
+	}
+	return c
+}
+func (c *SupportBundleController) watchSupportBundleCollections() {
+	klog.Info("Starting watch for SupportBundleCollections")
+	antreaClient, err := c.antreaClientGetter.GetAntreaClient()
+	if err != nil {
+		klog.Warningf("Failed to get antrea client: %v", err)
```

Comment:
new log shouldn't add more non-structured logging. It could use `ErrorS` if this is not expected and need user attention strongly and use `InfoS` if this is normal.